pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeConsumptionRateManager.java

```diff
@@ -0,0 +1,127 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This class is responsible for creating realtime consumption rate limiters. If rate limiter is used for
+ * multi-partition topics, the provided rate in StreamConfig is divided by partition count. This class leverages a
+ * cache for storing partition count for different topics as retrieving partition count from Kafka is a bit expensive
+ * and also the same count will be used of all partition consumers of the same topic.
+ */
+public class RealtimeConsumptionRateManager {
+  private static final Logger LOGGER = LoggerFactory.getLogger(RealtimeConsumptionRateManager.class);
+  private static final RealtimeConsumptionRateManager INSTANCE = new RealtimeConsumptionRateManager();
+  private static final int CACHE_ENTRY_EXPIRATION_TIME_IN_MINUTES = 10;
+
+  private final LoadingCache<StreamConfig, Integer> _streamConfigToTopicPartitionCountMap = buildCache();
+  private boolean _isThrottlingAllowed = false;
+
+  private RealtimeConsumptionRateManager() {
+  }
+
+  public static RealtimeConsumptionRateManager getInstance() {
+    return INSTANCE;
+  }
+
+  public void enableThrottling() {
+    _isThrottlingAllowed = true;
+  }
+
+  public ConsumptionRateLimiter createRateLimiterForSinglePartitionTopic(StreamConfig streamConfig) {
+    return createRateLimiter(streamConfig, false);
+  }
+
+  public ConsumptionRateLimiter createRateLimiterForMultiPartitionTopic(StreamConfig streamConfig) {
+    return createRateLimiter(streamConfig, true);
+  }
+
+  private ConsumptionRateLimiter createRateLimiter(StreamConfig streamConfig, boolean multiPartitionTopic) {
+    if (!streamConfig.getTopicConsumptionRateLimit().isPresent()) {
+      return NOOP_RATE_LIMITER;
+    }
+    int partitionCount = 1;
+    if (multiPartitionTopic) {
+      try {
+        partitionCount = _streamConfigToTopicPartitionCountMap.get(streamConfig);
+      } catch (ExecutionException e) {
+        // Exception here means that for some reason, partition count cannot be retrieved from Kafka broker!
+        throw new RuntimeException(e);
+      }
+    }
+    double topicRateLimit = streamConfig.getTopicConsumptionRateLimit().get();
+    double partitionRateLimit = topicRateLimit / partitionCount;
+    LOGGER.info("A rate limiter is setup for topic {} with rate limit: {} (topic rate limit: {}, partition count: {})",
+        streamConfig.getTopicName(), partitionRateLimit, topicRateLimit, partitionCount);
+    return new RateLimiterImpl(partitionRateLimit);
+  }
+
+  private LoadingCache<StreamConfig, Integer> buildCache() {
+    return CacheBuilder.newBuilder().expireAfterWrite(CACHE_ENTRY_EXPIRATION_TIME_IN_MINUTES, TimeUnit.MINUTES)
+        .build(new CacheLoader<StreamConfig, Integer>() {
+          @Override
+          public Integer load(StreamConfig streamConfig)
+              throws Exception {
+            String clientId = streamConfig.getTopicName() + "-consumption.rate.manager";
```

Comment:
Since we will have only one instance of this cache per host, I suggest we use the same client ID for all topics. Some streams start to emit metrics for each new client, and this can cause number of metrics to increase (although not by much).

Also, I am assuming CacehBuilder returns something like a concurrentHash

Reply:
This helps with debugging in case something goes wrong with metadata provider for one topic, but the other topics are fine. Also, as you mentioned, number metrics don't increase by much because for each topic, only one metadata provider is created and that's when the segment starts consuming and we immediately close the metadata provider after the partition count is fetched.